src/serac/physics/utilities/variational_form/weak_form.hpp

```diff
@@ -0,0 +1,472 @@
+   * @param[in] integrand The user-provided quadrature function, see @p Integral
+   * @param[in] domain The domain on which to evaluate the integral
+   * @note The @p Dimension parameters are used to assist in the deduction of the @a geometry_dim
+   * and @a spatial_dim template parameter
+   */
+  template <int geometry_dim, int spatial_dim, typename lambda>
+  void AddIntegral(Dimension<geometry_dim>, Dimension<spatial_dim>, lambda&& integrand, mfem::Mesh& domain)
+  {
+    if constexpr (geometry_dim == spatial_dim) {
+      auto num_elements = domain.GetNE();
+      SLIC_ERROR_IF(num_elements == 0, "Mesh has no elements");
+
+      auto dim = domain.Dimension();
+      for (int e = 0; e < num_elements; e++) {
+        SLIC_ERROR_IF(domain.GetElementType(e) != supported_types[dim], "Mesh contains unsupported element type");
+      }
+
+      const mfem::FiniteElement&   el = *test_space_->GetFE(0);
+      const mfem::IntegrationRule& ir = mfem::IntRules.Get(el.GetGeomType(), el.GetOrder() * 2);
+
+      constexpr auto flags = mfem::GeometricFactors::COORDINATES | mfem::GeometricFactors::JACOBIANS;
+      auto           geom  = domain.GetGeometricFactors(ir, flags);
+      domain_integrals_.emplace_back(num_elements, geom->J, geom->X, Dimension<geometry_dim>{},
+                                     Dimension<spatial_dim>{}, integrand);
+      return;
+    } else if constexpr ((geometry_dim + 1) == spatial_dim) {
+      // TODO: fix mfem::FaceGeometricFactors
+      constexpr bool always_false = (geometry_dim == spatial_dim);
+      static_assert(always_false,
+                    "unsupported integral dimensionality due to unimplemented features in mfem::FaceGeometricFactors");
+
+      auto num_boundary_elements = domain.GetNBE();
+      SLIC_ERROR_IF(num_boundary_elements == 0, "Mesh has no boundary elements");
+
+      auto dim = domain.Dimension();
+      SLIC_ERROR_IF(dim != spatial_dim, "Error: invalid mesh dimension for integral");
+      for (int e = 0; e < num_boundary_elements; e++) {
+        SLIC_ERROR_IF(domain.GetBdrElementType(e) != supported_types[dim - 1],
+                      "Mesh contains unsupported element type");
+      }
+
+      const mfem::FiniteElement&   el = *test_space_->GetBE(0);
+      const mfem::IntegrationRule& ir = mfem::IntRules.Get(el.GetGeomType(), el.GetOrder() * 2);
+      constexpr auto flags            = mfem::FaceGeometricFactors::COORDINATES | mfem::FaceGeometricFactors::JACOBIANS;
+
+      // despite what their documentation says, mfem doesn't actually support the JACOBIANS flag.
+      // this is currently a dealbreaker, as we need this information to do any calculations
+      auto geom = domain.GetFaceGeometricFactors(ir, flags, mfem::FaceType::Boundary);
+      boundary_integrals_.emplace_back(num_boundary_elements, geom->J, geom->X, Dimension<geometry_dim>{},
+                                       Dimension<spatial_dim>{}, integrand);
+      return;
+    } else if constexpr (true) {
```

Comment:
Why does a plain else not work?

Reply:
Good question! There's been some compiler warnings on NVCC related to `else`s corresponding to `if constexpr` expressions, but this doesn't look like one of those. I'll take another look.